### ISSUE TITLE
Question: Radio inputs with the same name attribute value must be part of a group (false positive?)

### ISSUE THREAD

**Issue by mdgbayly** [dequelabs/axe-core#176]
https://dequeuniversity.com/rules/axe/1.1/radiogroup

We have some markup with nested radio buttons that are generating the above error using the Chrome extension.  The only way we have found to resolve it is to add an identical aria-labelledby attribute to each radio button in the group referencing a common label.  However, our belief is that this does not give a good experience as we would like to include more contextual information for each radio button by using an aria-label or unique aria-labelledby attribute for each radio.

The example that demonstrates our issue is a slightly modified version of this WCAG role="radiogroup" example, which has been modified to include additional input type="text" fields in each sub group which matches our requirements.

https://www.w3.org/WAI/WCAG20/Techniques/working-examples/ARIA17/grouping-roles-example.html

The first version shown is the WCAG example with input type="text" fields added. It is using span's with role="radio" and interestingly this **passes** the aXe validation.

```
<div role="radiogroup" aria-labelledby="alert2">
    <p id="alert2">Send an alert when a charge exceeds $ 250</p>
    <div>
      <input type="text" name="text1" aria-label="text1">
      <span role="radio" aria-labelledby="a2r1" name="a2radio" tabindex="0" aria-checked="false"></span>
      <span id="a2r1">Yes</span>
    </div>

    <div>
      <input type="text" name="text2" aria-label="text2">
      <span role="radio" aria-labelledby="a2r2" name="a2radio" tabindex="-1" aria-checked="false"></span>
      <span id="a2r2">No</span>
    </div>
</div>
```

However, if you change the example to use actual radio buttons which more closely matches our desired markup, it fails with the error listed in the title. If you remove the additional input type="text fields, then it successfully passes aXe validation.

```
<div role="radiogroup" aria-labelledby="alert2">
    <p id="alert2">Send an alert when a charge exceeds $ 250</p>
    <div>
      <input type="text" name="text1" aria-label="text1">
      <input type="radio" aria-labelledby="a2r1" name="a2radio" tabindex="0" aria-checked="false"></input>
      <span id="a2r1">Yes</span>
    </div>

    <div>
      <input type="text" name="text2" aria-label="text2">
      <input type="radio" aria-labelledby="a2r2" name="a2radio" tabindex="-1" aria-checked="false"></input>
      <span id="a2r2">No</span>
    </div>
</div>
```

I understand that this markup might be questionable, given the nesting and different types of inputs in the role="radiogroup" grouping element. I've seen other scenarios where aXe flags an error in situations like that due to unassociated inputs or something.  However, NVDA recognises the radio button grouping still in the second example.  Also it seems inconsistent at least that aXe does not treat the above two scenarios the same?

**Comment by dylanb** [comment]
Are you saying that this fails:
```
<div role="radiogroup" aria-labelledby="alert2">
    <p id="alert2">Send an alert when a charge exceeds $ 250</p>
    <div>
      <input type="text" name="text1" aria-label="text1">
      <input type="radio" aria-labelledby="a2r1" name="a2radio" tabindex="0" aria-checked="false"></input>
      <span id="a2r1">Yes</span>
    </div>

    <div>
      <input type="text" name="text2" aria-label="text2">
      <input type="radio" aria-labelledby="a2r2" name="a2radio" tabindex="-1" aria-checked="false"></input>
      <span id="a2r2">No</span>
    </div>
</div>
```

but this passes?

```
<div role="radiogroup" aria-labelledby="alert2">
    <p id="alert2">Send an alert when a charge exceeds $ 250</p>
    <div>
      <input type="radio" aria-labelledby="a2r1" name="a2radio" tabindex="0" aria-checked="false"></input>
      <span id="a2r1">Yes</span>
    </div>

    <div>
      <input type="radio" aria-labelledby="a2r2" name="a2radio" tabindex="-1" aria-checked="false"></input>
      <span id="a2r2">No</span>
    </div>
</div>
```

**Comment by mdgbayly** [comment]
Yes

Here is a jsfiddle ( if you can ignore all the other errors caused by the fiddle)

With inputs: https://jsfiddle.net/3okf871L/

Without inputs: https://jsfiddle.net/3okf871L/1/

**Comment by WilcoFiers** [comment]
It looks to me like the difference here comes from a bug in axe. It should have not just looked at native input elements but at elements with role=radiobutton. Another issue here is that the check that looks at this also only looks at other native input elements. Both of these doen't seem to be the original intent of this rule.

I do think we may have constrained these radiogroup rules a bit too much. We may change this to a best practice. What we could do (speculating here) is check radiobuttons with the same label text do have a different group name. This way we avoid failing self explanatory radiobutton labels.

**Comment by dylanb** [comment]
ok, so what we are going to do here is make it fail both cases.

We also are going to demote this issue to a best practice because after further analysis we believe that, while better, there are definitely instances where no group label is acceptable.

**Comment by mdgbayly** [comment]
OK - I'm sure you guys are right as you're way more clued into this than I am, but I'm still interested to understand why these are not valid from an accessibility/semantic HTML point of view.

I see that aXe also flags the following with the same error.
It gives the following error summary:
```
Summary:

Fix any of the following:

All elements with the name "a2radio" do not reference the same element with aria-labelledby
Fieldset contains unrelated inputs
```
https://jsfiddle.net/3okf871L/2/

```
<div>
    <p id="alert2">Send an alert when a charge exceeds $ 250</p>
    <fieldset>
      <legend>
        Mixed inputs
      </legend>
      <div>
        <input type="text" name="text1" aria-label="text1">
        <input type="radio" aria-labelledby="a2r1" name="a2radio" tabindex="0" aria-checked="false"></input>
        <span id="a2r1">Yes</span>
      </div>

      <div>
        <input type="text" name="text2" aria-label="text2">
        <input type="radio" aria-labelledby="a2r2" name="a2radio" tabindex="-1" aria-checked="false"></input>
        <span id="a2r2">No</span>
      </div>
    </fieldset>
</div>
```
Yet it seems valid to have 'unrelated inputs' in a fieldset according to the HTML5 spec:

https://dev.w3.org/html5/spec-preview/the-fieldset-element.html

```
<fieldset>
 <legend>Display</legend>
 <p><label><input type=radio name=c value=0 checked> Black on White</label>
 <p><label><input type=radio name=c value=1> White on Black</label>
 <p><label><input type=checkbox name=g> Use grayscale</label>
 <p><label>Enhance contrast <input type=range name=e list=contrast min=0 max=100 value=0 step=1></label>
 <datalist id=contrast>
  <option label=Normal value=0>
  <option label=Maximum value=100>
 </datalist>
</fieldset>
```
Can you point me to something in WCAG or similar that you are using as the basis for your 
```
Fieldset contains unrelated inputs
```

I'm sure you're right - just trying to educate myself.
Thanks

**Comment by dylanb** [comment]
@mdgbayly on further reflection, we have decided that this portion of the rule is generating false positives (your example is in fact a false positives) and as a result, we will remove it.

Just for the record, here is why that code is currently there: having a radio group like that with the same name, makes the browser interpret certain keyboard commands (up and down arrow) to move the selection and the keyboard focus between the items in the group. If there are a lot of other elements interspersed within that group, this can lead to difficulty understanding where in the form you are (as a screen reader user for example).

Also, in the case of checkbox groupings with much stuff interspersed, the difficulty can be in understanding how all the fields relate to one another.

In your case (as long as you add labels for the text inputs) this does not happen because there is not a lot of stuff in-between.

**Comment by mdgbayly** [comment]
Ok thanks for the feedback.

:bow: great work you guys are doing by the way.  We take accessibility very seriously where we work, and the tools you guys are creating are making things so much easier.

**Comment by dylanb** [comment]
Awesome! Your platform is very important for people with disabilities to be able to get a good education!

go #a11y!

**Comment by gerardkcohen** [comment]
Another case of this error, radio buttons in a table as a single select pattern.

Ex:

```
        <table>
            <caption>Select a Product</caption>
            <thead>
                <tr>
                    <th scope="col">Select</th>
                    <th scope="col">Product</th>
                    <th scope="col">Description</th>
                    <th scope="col">Cost</th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <th scope="row">Basic Web Server</th>
                    <td>
                        <label>Select Basic
                            <input type="radio" name="product" value="basic" />
                        </label>
                    </td>
                    <td>Basic server to host your basic site</td>
                    <td>$199/yr</td>
                </tr>
                <tr>
                    <th scope="row">Awesome Web Server</th>
                    <td>
                        <label>Select Awesome
                            <input type="radio" name="product" value="awesome" />
                        </label>
                    </td>
                    <td>Awesome server to host your awesome site</td>
                    <td>$299/yr</td>
                </tr>
                <tr>
                    <th scope="row">SSL Cert</th>
                    <td>
                        <label>Select SSL Cert
                            <input type="radio" name="product" value="ssl" />
                        </label>
                    </td>
                    <td>Make sure your site is secure</td>
                    <td>$99/yr</td>
                </tr>
            </tbody>
        </table>
```

**Comment by dylanb** [comment]
@gerardkcohen can you move your example to a separate issue - it is complicated enough that it deserves separate discussion.

**Comment by gerardkcohen** [comment]
@dylanb Thanks! Issue #306

**Comment by cheapsteak** [comment]
This bug still seems to occur in current version of axe

**Comment by WilcoFiers** [comment]
@cheapsteak It's still an open issue. This one has been on our radar for some time now but we haven't managed to get to it yet, so we'd very much welcome a pull request.

**Comment by slightlytyler** [comment]
<img width="1329" alt="ffa44136-5d3c-11e9-97f5-cc817513105d (1)" src="https://user-images.githubusercontent.com/6819171/56165823-cd88db00-5f88-11e9-8de9-53d7eb6247de.png">

I've been experiencing this same problem in my PR to add storybook to our project

https://git.musta.ch/airbnb/airbnb/pull/174508

if anyone needs a repro checkout out that PR and run the `city-registration` storybook then choose `BlueprintQuestion` => `Radio question`

**Comment by fabdotnet** [comment]
I had the same issue and ended up solving it by replacing the div group with a fieldset + legend:

`<fieldset>
      <legend>I am legend</legend>
<div>
        <input type="radio" id="yes" name="group-name"></input>
        <label for="yes">Yes</span>
      </div>

      <div>
        <input type="radio" id="no" name="group-name"></input>
        <label for="no">No</span>
      </div>
</fieldset>`

**Comment by jeankaplansky** [comment]
@jeankaplansky : See above - Need to deprecate these two rules in 3.4 (but not delete).

**Comment by jeankaplansky** [comment]
Rule deprecated in 3.4. Rule help unpublished for 3.4
